### DISCUSSION TITLE
Recursive structures with cue

### DISCUSSION THREAD

**Discussion by pipex** [cuelang/cue#594] [Q&A] [answered]
Hi everyone,

I'm experimenting with cue to try to specify some hierarchical configuration structures, however I'm having troubles with recursivity in cue. I'm sure there is something I'm misunderstanding though.

This is the type of definition I'm doing

```
// test.cue
#Base: {
  key?: string
  ...
}

#Single: #Base & {
  value?: string
}

#Elem: #Single | #Compound

#Compound: #Base & {
  elems: [Key=string]: #Elem & {
    key: Key
  }
}

test: #Compound & {
  elems: one: {} 
  elems: two: {
    elems: twoA: {}
    elems: twoB: {}
  } 
}
```

But when I do `cue export test.cue` I get (in 0.2.2)

```
test.elems.one: cannot convert incomplete value "(C{key: (string & Key), value?: string} | C{key: (string & Key), elems: {[]: (Key: string)->(#Elem & C{key: Key}), }})" to JSON:
    ./test.cue:19:7
    ./test.cue:14:24
    ./test.cue:20:15
```

And in (0.3.0-alpha4)

```
test.elems.one: cannot convert incomplete value "(struct){ |((#struct){\n    key: (string){ \"one\" }\n  }, (#struct){\n    key: (string){ \"one\" }\n    elems: (#struct){\n    }\n  }) }" to JSON
```

When i would expect 

```
{
    "test": {
        "elems": {
            "one": {
                "key": "one"
            },
            "two": {
                "key": "two",
                "elems": {
                    "twoA": {
                        "key": "twoA",
                    },
                    "twoB": {
                        "key": "twoB",
                    }
                }
            }
        }
    }
}
```

When I replace the definition of  `#Compound` with

```
#Compound: #Base & {
  elems: [Key=string]: #Single & {
    key: Key
  }
}
```

I get 

```
{
    "test": {
        "elems": {
            "one": {
                "key": "one"
            },
            "two": {
                "key": "two",
                "elems": {
                    "twoA": {},
                    "twoB": {}
                }
            }
        }
    }
}
```

Which is closer.

So I guess the question is, what would be the proper way to do this kind of definition in cue? I'm kind of a newbie to the language, so not sure if I'm doing something wrong or really this is not something that is supported.

**Comment by mpvl** [comment, marked as answer]
[using v0.3.0-* for all of this]

The error is a lot clearer if you export to cue (JSON errors need improvement, clearly):
```
cue export --out cue
```
```
test.elems.one: incomplete value {key:"one"} | {key:"one",elems:{}}
test.elems.two: incomplete value {elems:{twoA:{},twoB:{}},key:"two"} | {elems:{twoA:{key:"twoA"} | {key:"twoA",elems:{}},twoB:{key:"twoB"} | {key:"twoB",elems:{}}},key:"two"}
```
The problem is that in `#Elem` both structs of the disjunction ("sum type") allow `key`. So with just the line
```
 elems: [Key=string]: #Elem & {
    key: Key
  }
```
it still cannot be decided which of the two structs should end up in the result.

At one point the spec stated that the evaluator is allowed to pick the most specific one. So in your case, that means it would have picked `#Single`.  But this is expensive, sometimes not computable, and not even always what users want, so that is no longer supported.

Instead, you can add a default marker to resolve the ambiguity:
```
#Elem: *#Single | #Compound
```
This will cause the #Single to be picked in case both match.

This is probably not what you want, though. It would always match, as `#Base` contains an `...` and thus allows any field.

What you probably want is this:
```
#Base: {
	key?: string
}

#Single: {
	#Base

	value?: string
}

#Elem: #Single | #Compound

#Compound: {
	#Base

	elems: [Key=string]: #Elem & {
		key: Key
	}
}

test: #Compound & {
	elems: one: {value: "a"} // added value to pick #Single
	elems: two: {
		elems: twoA: {value: "b"} // added value to pick #Single
		elems: twoB: {value: "c"} // added value to pick #Single
	}
}
```
Here, instead of using `...`, I've "embedded" `#Base` within the two derivative types. Note that `#Base` is "closed" and only allows the field `key` now. Embedding allows adding fields to a closed type to create a new type. The result is also closed. This is syntactically and semantically very close to Go's struct embedding. It actually works somewhat better in CUE, as overlapping fields can be meaningfully defined in CUE.

Note that I removed the default marker and explicitly chose `#Single` for each of the fields to ensure the result is fully resolved. This would not have worked with the `...` as both variants would still have matched.

The result is now:
```
$ cue export foo.cue 
{
    "test": {
        "elems": {
            "one": {
                "value": "a",
                "key": "one"
            },
            "two": {
                "key": "two",
                "elems": {
                    "twoA": {
                        "value": "b",
                        "key": "twoA"
                    },
                    "twoB": {
                        "value": "c",
                        "key": "twoB"
                    }
                }
            }
        }
    }
}
```

**Reply by pipex** [reply]
Wow, thanks for the detailed explanation. 

That makes it super clear what was missing in my definition. Embedding looks like the way to go then. Thanks again!

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/594.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
